numpy/core/src/npymath/npy_math_complex.c.src

```diff
@@ -445,28 +445,26 @@ npy_cpow@c@ (@ctype@ a, @ctype@ b)
     @type@ bi = npy_cimag@c@(b);
     @ctype@ r;
 
-    if (br == 0. && bi == 0.) {
-        return npy_cpack@c@(1., 0.);
-    }
-    if (ar == 0. && ai == 0.) {
-        if (br > 0 && bi == 0) {
-            return npy_cpack@c@(0., 0.);
+    /*Checking if in a^b, if b is zero.
+    If a is not zero then by definition of logarithm a^0 is one.
+    If a is also zero then as per IEEE 0^0 is best defined as 1.*/
+    if(br == 0. && bi == 0.){
+        return npy_cpack@c@(1.,0.);
+    }
+    /*Here it is already tested that for a^b, b is not zero.
+    So we are checking behaviour of a.*/
+    else if(ar == 0. && ai == 0.){
+        /*Checking if real part of power is greater than zero then the result is zero.
+        If not the result is undefined as it blows up or oscillates.*/
+        if(br > 0){
+            return npy_cpack@c@(0.,0.);
         }
-        else {
-            volatile @type@ tmp = NPY_INFINITY@C@;
-            /*
-             * NB: there are four complex zeros; c0 = (+-0, +-0), so that
-             * unlike for reals, c0**p, with `p` negative is in general
-             * ill-defined.
```

Comment:
Ok this NB might point to something I had missed, after removing all our custom handling (lets ignore the warnings there, they may be glibc warning too much):
```
In [39]: np.power(0-0j, 1-1j)
<ipython-input-39-4592e7f77c06>:1: RuntimeWarning: divide by zero encountered in power
  np.power(0-0j, 1-1j)
Out[39]: 0j

In [40]: np.power(0-0j, 1+1j)
<ipython-input-40-a7fd9d0e2031>:1: RuntimeWarning: divide by zero encountered in power
  np.power(0-0j, 1+1j)
Out[40]: -0j
```
i.e. the result actually (subtly) depends on the imaginary part in the branch above.  I expect that is the reason why we had the `bi == 0` check there previously.

Reply:
@seberg The check for bi==0 is already in place with br==0 just like before. (Line 448 in original file and Line 451 in pull request file)

Reply:
No, there is no check for `bi == 0` when `br != 0`.  If `br > 0` and `bi != 0` the sign of the result can still change, as my example shows I think. Even though the value is always 0.

The old NB comment tries to point that out.  I don't know what should happen exactly, but if we delete that comment (confusing as it is), we should either get the signs right or replace it with an arguement that it is OK to get the sign wrong because of XY.

Reply:
@seberg I need to check if it is ok to get signs wrong or what can be done to make the signs right. I might need some time though.

Reply:
Thanks, its complicated :(, I wish it was easier to get this stuff right...